### ISSUE TITLE
Add support for custom SessionIndex

### ISSUE THREAD

**Issue by pokej6** [simplesamlphp/simplesamlphp#1399]
### Overview
Today someone can set up simplesamlphp as an identity provider that still relies on the user logging in to an external authentication system. This is done by extending the `Source` base class for an authentication source. That external system is essentially the identity provider for simplesamlphp in this case and is likely also managing its own session. If that external application is controlling centralized session management for a larger system one might want that external application to be able to kill saml sessions and/or trigger LogoutRequests on the back channel. This would most-easily be achieved by using saml constructs such as SessionIndex; when a user logs in, the external application can provide a token of some sort that is sent in the SamlAssertion to service providers so that later during a logout the request cascades to all applications managed by that central system including saml-enabled applications. Unfortunately today if you follow the workflow there's no easy way to override the SessionIndex that gets sent in an assertion from simplesamlphp.

`Source::loginCompleted` eventually calls its `$return` which is `IdP::postAuth` which has a `ProcessingChain` that gets invoked. This chain has a bunch of filters that operate on the request. It's possible that this could capture an incoming "session index" and then save it in the state for use later. Eventually you get to `SAML2::sendResponse` which calls `buildAssertion` with the `state`. Unfortunately when building the assertion, this is the line that handles the SessionIndex:
```php
$a->setSessionIndex(Utils\Random::generateID());
```
This does not leave any method of controlling this value in a generated assertion.

### Proposed Solutions
It would be great if this line could do one of the following:

1. Check the state for some value set that it should use as the SessionIndex
1. Create a SessionIndexProvider interface that gets used here so that someone could provide their own implementation for how a SessionIndex is created. It may be beneficial to have this method be passed the state object.

### What I've Tried
The code in question is pretty static. There are a few methods I've considered for getting around this but I'm not really fond of any of them.

1. Monkey patch `Utils\Random::generateID` to have it read from php superglobals in order to get other request information when generating a session index. Clearly monkey patching anything is undesirable.
1. Rewriting a significant portion of the SAML flow. A lot of the flow is handled through static methods so this seems like a lot of work for little gain. I'd rather not.
1. Add a custom saml attribute that essentially is SessionIndex and make sure our consumers know to treat it that way. This isn't desirable since SessionIndex is already part of the saml specification. It'd be great to just use it.

### Additional context
I'm currently toying with simplesamlphp 1.18.x. I see 1.19 is out for RC but would love if this change, if worked on, would make it into the 1.18.x branch so that I didn't need to start looking at updating my PHP version.

**Comment by pokej6** [comment]
Bumping this to make sure someone has at least seen it. 

@tvdijen Is there anything else you need me to add to this?

**Comment by tvdijen** [comment]
I've call in backup, because I'm not sure about this in relation to the specification... It's pretty vague on what the intended purpose of the SessionIndex-flag is

**Comment by jaimeperez** [comment]
Hi @pokej6!

I see your point, but I think the approach to try to use `SessionIndex` for it might be wrong (or at least suboptimal).

SimpleSAMLphp requires cookies, and we keep a session for every user. Similarly, upstream systems where we delegate authentication to might also support sessions with cookies, in which case you would just need to tell them (somehow, that's not in scope since we're not talking SAML IdPs here) that you want to log the user out, and then, upon redirecting there, that _IdP_ would get the session cookies and be able to log the user out.

Now, if that external authentication source of some kind does *not* support cookies, like in the case you are describing of a back channel connection, you will indeed need to pass it some identifier so that the system knows which session to kill. And here is where I think your approach is wrong. Since you have a session yourself in SimpleSAMLphp, you can store data in there if you like. That data can be the very same _token_ you are willing to pass along in the `SessionIndex`. So when your auth source is called to log the user out, all it'll need to do is fetch the _token_ and trigger a back-channel request to the remote system with it.

The reason why I believe you should not use `SessionIndex` for this is privacy. Even if you send just a targeted identifier to two different services, they would then be able to correlate the `SessionIndex`'s they have received to track the user (and potentially link targeted identifiers, so that the correlation becomes permanent). You could of course make the session indexes targeted as well to get rid of that, based on the original token + service-specific information. However, in order for services not to be able to find out the original token, you would need a one-way (hash) function, and then you won't be able to recover the token yourself when logging out, so you'll need to keep a mapping table somewhere. Meaning you will save the original token anyway. So in that case, the simplest solution is to just keep that token in the session, then use it when forcing the upstream system to log the user out.

Does that make sense, or did I misunderstood something?

**Comment by pokej6** [comment]
The `SessionIndex` is used as an identifier for the user's session. A SAML Service Provider should associate any session it creates as a result of a received SAML assertion with the `SessionIndex` in the SAMLResponse. Later when a LogoutRequest is sent (either by the IdP or an SP) it can contain a `SessionIndex` to specify certain sessions that should be logged out by the SP or IdP. 

This process usually starts on the front channel with a user logging out of a service that they have been logged in with. The SP sends a LogoutRequest to the IdP. At that point the IdP can issue LogoutRequests on the back channel to all other SPs with which that user is known to be logged in (Consider this diagram https://www.identityserver.com/media/1182/saml-slo.png). There is no leak of actual session information here because the only thing that gets passed around is the SessionIndex.

A SessionIndex is not necessarily unique to a user. Theoretically an IdP can use a singular SessionIndex to represent a timeframe and then log out all sessions tied to that timeframe or something like that (this is contrived. I haven't really seen this in the wild).

**Comment by tvdijen** [comment]
Aside from any of this... Why on earth do you rely on anything related to SLO when you know it's not safe?

**Comment by pokej6** [comment]
> when you know it's not safe

Can you explain what you mean by this? SLO is part of the SAML specification and is implemented by other SAML IdP solutions such as Okta and PingFederate.

**Comment by tvdijen** [comment]
Yes, but SLO is unsafe, because you send your SLO, but you are never certain to have the reply back..  The person using it things he logged out, but you never really know if you have..

**Comment by pokej6** [comment]
> you send your SLO, but you are never certain to have the reply back
That is the entire point of the asynchronous binding for the SLO profile but I never mentioned whether I was doing synchronous or asynchronous binding. It's not really relevant. 

> The person using it thinks he logged out, but you never really know if you have
I'm still not sure why that makes SLO "unsafe". You never really **know** if you're logged out of anything unless you manually keep track of all services with which you've authenticated, navigate to each one of them, attempt to log out, and then try using your session again afterwards. Is this really how you expect a user to log out of the services they have accessed?

You _might_ be thinking it's unsafe in a scenario where SLO is completely handled on the frontend. In this workflow the IdP, upon receiving your SLO request, directs you, through your user agent (browser), to each service you've logged in to with an SLO request. There are a couple breakdowns in this workflow. One is that any SP in the chain could error, be down for maintenance, or simply decide not to respond in an expected manner. Another is that the user could just interrupt the process amidst all the redirects. In both those of cases the IdP would never be able finish the logout process. However if this process is handled purely with server-to-server communication then the IdP is fully aware of any failures and can proceed anyway. (See Figure 3 of [4.4.2 Profile Overview](https://docs.oasis-open.org/security/saml/v2.0/saml-profiles-2.0-os.pdf). The User Agent is not required here when the IdP directly communicates with the SP but it can be used if everything goes through the browser)

Regardless of opinions of SLO, the bottom line of my enhancement request is to allow for more control over the SessionIndex sent in a SAML Assertion as it is presently not customizable.

**Comment by pokej6** [comment]
Is this going to be addressed? I took a look at old commits and the original simplesamlphp implementation allowed for custom SessionIndex via setting it in the Session. The reason it was removed was not due to security but to make it so that each Service Provider could have their own SessionIndex (https://github.com/simplesamlphp/simplesamlphp/commit/2ff5de5ef9fbc638d418a5489d179c1ba187073c)
